src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/IO/TdsWriter.cs

```diff
@@ -0,0 +1,187 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Buffers.Binary;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Data.SqlClient;
+
+namespace Microsoft.Data.SqlClientX.IO
+{
+    internal class TdsWriter
+    {
+        private readonly TdsWriteStream _writeStream;
+
+        // Header length constants
+        private const int OutputHeaderLen = TdsEnums.HEADER_LEN;
+
+        // Out buffer variables
+        private readonly byte[] _outBuff; // internal write buffer - initialize on login
+        private int _outBytesUsed = OutputHeaderLen; // number of bytes used in internal write buffer - initialize past header
+
+        /// <summary>
+        /// Instantiate TdsWriter with a TdsWriteStream
+        /// </summary>
+        /// <param name="stream">A Tds Writer stream to work with</param>
+        /// <exception cref="ArgumentNullException"></exception>
+        public TdsWriter(TdsWriteStream stream)
+        {
+            _writeStream = stream ?? throw new ArgumentNullException(nameof(stream));
+            _outBuff = new byte[OutputHeaderLen]; // Adjust the buffer size as needed
+        }
+
+        #region Public APIs
+
+        /// <summary>
+        /// Writes short value to out buffer
+        /// </summary>
+        /// <param name="v">Value to write</param>
+        /// <param name="isAsync">Is Async caller method</param>
+        /// <param name="ct">Cancellation token</param>
+        public async Task WriteShortAsync(short v, bool isAsync, CancellationToken ct)
+        {
+            await EnsureBufferSpaceAsync(2, isAsync, ct).ConfigureAwait(false);
```

Comment:
Here and for everything, there is no need to ensure buffer space. The buffer is only maintained by the stream. 
The job of the caller is to write bytes to it .